book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,497 @@
+        self.transparent - [self.sprout + self.sapling + self.orchard].sum()
+    }
+}
+
+impl Add for Result<ValueBalance<C>>
+where
+    C: Constraint,
+{
+
+}
+
+impl Sub for Result<ValueBalance<C>>
+where
+    C: Constraint,
+{
+
+}
+
+impl AddAssign for Result<ValueBalance<C>>
+where
+    C: Constraint,
+{
+
+}
+
+impl SubAssign for Result<ValueBalance<C>>
+where
+    C: Constraint,
+{
+
+}
+
+impl Sum for Result<ValueBalance<C>>
+where
+    C: Constraint,
+{
+
+}
+
+impl Default for ValueBalance<C>
+where
+    C: Constraint,
+{
+
+}
+```
+
+### Create a method in `Transaction` that returns `ValueBalance<NegativeAllowed>` for the transaction
+
+We first add `value_balance()` methods in all the modules we need and use them to get the value balance for the whole transaction.
+
+#### Create a method in `Input` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transparent.rs`.
+- Method need `utxos`, this information is available in `verify_transparent_inputs_and_outputs`.
+- If the utxos are not available in the block or state, verification will timeout and return an error
+
+```rust
+impl Input {
+    fn value_balance(&self, utxos: &HashMap<OutPoint, Utxo>) -> ValueBalance<NegativeAllowed> {
+
+    }
+}
+```
+
+#### Create a method in `Output` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transparent.rs`.
+
+```rust
+impl Output {
+    fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+    }
+}
+```
+
+#### Create a method in `JoinSplitData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/joinsplit.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+}
+```
+
+#### Create a method in `sapling::ShieldedData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/sapling/shielded_data.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+}
+```
+
+#### Create a method in `orchard::ShieldedData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/orchard/shielded_data.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+}
+```
+
+#### Create the `Transaction` method
+
+- Method location: `zebra-chain/src/transaction.rs`
+- Method will use all the `value_balances()` we created until now.
+- Method will check the remaining transaction value consensus rule.
+
+```rust
+/// utxos must contain the utxos of every input in the transaction,
+/// including UTXOs created by earlier transactions in this block.
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> Result<ValueBalance<NegativeAllowed>, Err> {
+    ..
+    // As soon as we have the value balance for all the pools
+    // check the remaining transaction value consensus rule:
+    value_balance.remaining_transaction_value()?;
```

Comment:
Efficiency / Consistency: This check will work, but we'll do it every time we access the `Transaction::value_balance`. Which might waste some CPU for transactions with lots of inputs and outputs.

We do similar checks in the transaction verifier:
https://github.com/ZcashFoundation/zebra/blob/f817df638b1ba8cf8c261c536a30599c805cf04c/zebra-consensus/src/transaction.rs#L168-L177

Can we move this check there?

Reply:
Moved at https://github.com/ZcashFoundation/zebra/pull/2430/commits/1974aab8d7c28ca3d50c2636d59cc4e14ef98b0d